src/main/java/org/elasticsearch/common/util/MultiDataPathUpgrader.java

```diff
@@ -0,0 +1,328 @@
+            mostUsableSpace = Math.max(mostUsableSpace, info.usableSpace + info.spaceUsedByShard);
+        }
+
+        if (mostUsableSpace < totalBytesUsedByShard) {
+            throw new IllegalStateException("Can't upgrade path available space: " + mostUsableSpace + " bytes required space: " + totalBytesUsedByShard + " bytes");
+        }
+        ShardFileInfo target = shardFileInfo[0];
+        if (leastUsableSpace >= (2 * totalBytesUsedByShard)) {
+            for (ShardFileInfo info : shardFileInfo) {
+                if (info.spaceUsedByShard > target.spaceUsedByShard) {
+                    target = info;
+                }
+            }
+        } else {
+            for (ShardFileInfo info : shardFileInfo) {
+                if (info.usableSpace > target.usableSpace) {
+                    target = info;
+                }
+            }
+        }
+        return new ShardPath(target.path, target.path, settings.get(IndexMetaData.SETTING_UUID), shard);
+    }
+
+    private ShardFileInfo[] getShardFileInfo(ShardId shard) throws IOException {
+        final Path[] paths = nodeEnvironment.shardPaths(shard); // custom data path doesn't need upgrading
+        final ShardFileInfo[] info = new ShardFileInfo[paths.length];
+        for (int i = 0; i < info.length; i++) {
+            Path path = paths[i];
+            final long usableSpace = getUsabelSpace(path);
+            info[i] = new ShardFileInfo(path, usableSpace, getSpaceUsedByShard(path));
+        }
+        return info;
+    }
+
+    protected long getSpaceUsedByShard(Path path) throws IOException {
+        final long[] spaceUsedByShard = new long[] {0};
+        if (Files.exists(path)) {
+            Files.walkFileTree(path, new FileVisitor<Path>() {
+                @Override
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    if (attrs.isRegularFile()) {
+                        spaceUsedByShard[0] += attrs.size();
+                    }
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+        }
+        return spaceUsedByShard[0];
+    }
+
+    protected long getUsabelSpace(Path path) throws IOException {
+        FileStore fileStore = Files.getFileStore(path);
```

Comment:
Hmm this method is buggy (@rmuir knows the details!); Lucene uses a hack workaround for spins() that I recently poached and put into NodeEnvironement.  Maybe factor that out somewhere and use it here too?